### PR TITLE
KAFKA-1194: Fixes for Windows

### PR DESCRIPTION
**Pull request by simplesteph**
Adapt changes from #4431 to existing codebase

Original contributors @nxmbriggs404 and @GeorgeCGV

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/AbstractIndex.scala [hunk 1]
```diff
@@ -316,8 +341,10 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
    * Forcefully free the buffer's mmap.
    */
   protected[log] def forceUnmap() {
-    try MappedByteBuffers.unmap(file.getAbsolutePath, mmap)
-    finally mmap = null // Accessing unmapped mmap crashes JVM by SEGV so we null it out to be safe
+    if (mmap != null) {
```
Comment (simplesteph) [inline core/src/main/scala/kafka/log/AbstractIndex.scala:362]: doing the NPE check helps fix a WARN in the log (see https://pastebin.com/NHeAcc6v). Unsure why `mmap` in the first place is null, probably this method might have been called twice? Anyway, I don't think it hurts to check for NPE in that case

```
[2018-09-03 00:16:04,902] INFO Updated PartitionLeaderEpoch. New: {epoch:0, offset:0}, Current: {epoch:-1, offset:-1} for Partition: test4-0. Cache now contains 0 entries. (kafka.server.epoch.LeaderEpochFileCache)
[2018-09-03 00:16:53,361] INFO Deleted log C:\tmp\kafka-logs\test3-0.d2d44094e2a74667b84ae6d843835a57-delete\00000000000000000000.log. (kafka.log.LogSegment)
[2018-09-03 00:16:53,361] INFO Deleted offset index C:\tmp\kafka-logs\test3-0.d2d44094e2a74667b84ae6d843835a57-delete\00000000000000000000.index. (kafka.log.LogSegment)
[2018-09-03 00:16:53,361] INFO Deleted time index C:\tmp\kafka-logs\test3-0.d2d44094e2a74667b84ae6d843835a57-delete\00000000000000000000.timeindex. (kafka.log.LogSegment)
[2018-09-03 00:16:53,375] INFO Deleted log for partition test3-0 in C:\tmp\kafka-logs\test3-0.d2d44094e2a74667b84ae6d843835a57-delete. (kafka.log.LogManager)
```

**Comment by simplesteph** [comment]
@lindong28 @ijuma this fixes KAFKA-1194 as per my tests on Windows.

Honestly I'm not sure which part of the code actually fixes it, and some of it actually breaks a few tests. I've just referenced previous PRs and fixed a NPE as well. 

I'm looking for your expertise to look at this and probably let me know what could be reverted to the original codebase, and what can be kept to fix 1194 cleanly and let the test suite pass

I appreciate your help on this

Cheers
Stephane

**Comment by simplesteph** [comment]
One build failed with one Kafka Streams test (NPE) which I believe is unrelated to this PR: https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/16248/testReport/

Reopened to trigger a new build

**Comment by simplesteph** [comment]
@ijuma @lindong28 this passes tests and fixes KAFKA-1194 as per my manual testing.

File: clients/src/main/java/org/apache/kafka/common/record/FileRecords.java [hunk 2]
```diff
@@ -438,16 +438,20 @@ private static FileChannel openChannel(File file,
...
+                return FileChannel.open(file.toPath(), StandardOpenOption.WRITE, StandardOpenOption.READ);
             } else {
-                RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw");
-                randomAccessFile.setLength(initFileSize);
-                return randomAccessFile.getChannel();
+                if (preallocate) {
+                    try (RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw")) {
+                        randomAccessFile.setLength(initFileSize);
+                    }
+                    return FileChannel.open(file.toPath(), StandardOpenOption.WRITE, StandardOpenOption.READ);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/FileRecords.java:459]: Why is this needed? This is worse than the previous approach as it opens, closes and reopens the file.

File: clients/src/main/java/org/apache/kafka/common/record/FileRecords.java [hunk 3]
```diff
@@ -438,16 +438,20 @@ private static FileChannel openChannel(File file,
...
+                } else {
+                    return FileChannel.open(file.toPath(), StandardOpenOption.WRITE, StandardOpenOption.READ, StandardOpenOption.CREATE);
+                }
             }
         } else {
-            return FileChannel.open(file.toPath());
+            return FileChannel.open(file.toPath(), StandardOpenOption.READ);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/FileRecords.java:465]: `READ` is the default on Linux, is it not on Windows? Is this change actually needed?

File: clients/src/main/java/org/apache/kafka/common/record/FileRecords.java [hunk 4]
```diff
@@ -438,16 +438,20 @@ private static FileChannel openChannel(File file,
                                            int initFileSize,
                                            boolean preallocate) throws IOException {
         if (mutable) {
-            if (fileAlreadyExists || !preallocate) {
-                return FileChannel.open(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.READ,
-                        StandardOpenOption.WRITE);
+            if (fileAlreadyExists) {
+                return FileChannel.open(file.toPath(), StandardOpenOption.WRITE, StandardOpenOption.READ);
             } else {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/FileRecords.java:454]: Instead of having an `else` with a nested `if/else`, we could keep it flat:

```scala
if (fileAlreadyExists)
...
else if (preallocate)
...
else
...
```

File: core/src/main/scala/kafka/log/AbstractIndex.scala [hunk 5]
```diff
@@ -202,8 +202,26 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
...
+            this.mmap = raf.getChannel.map(FileChannel.MapMode.READ_WRITE, 0, len)
+            this.mmap.position(position)
+          } finally {
+            CoreUtils.swallow(raf.close(), this)
+          }
+        }
+      } finally {
+        file = f
+      }
+    }
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractIndex.scala]: This code is pretty confusing. I suggest having a separate path for Windows instead of mixing the logic.

File: core/src/main/scala/kafka/log/AbstractIndex.scala [hunk 6]
```diff
@@ -228,7 +246,10 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
       // However, in some cases it can pause application threads(STW) for a long moment reading metadata from a physical disk.
       // To prevent this, we forcefully cleanup memory mapping within proper execution which never affects API responsiveness.
       // See https://issues.apache.org/jira/browse/KAFKA-4614 for the details.
-      safeForceUnmap()
+      CoreUtils.swallow(forceUnmap(), this)
+      // Accessing unmapped mmap crashes JVM by SEGV.
+      // Accessing it after this method called sounds like a bug but for safety, assign null and do not allow later access.
+      mmap = null
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractIndex.scala]: forceUnmap sets the `mmap` to `null` so not clear why we need this change at all. I would revert it and test again.

File: core/src/main/scala/kafka/log/AbstractIndex.scala (hunk 1 above)
Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractIndex.scala:362]: We shouldn't do this, we need to understand why we are calling it twice as that might be a bug.

File: core/src/main/scala/kafka/log/Log.scala [hunk 7]
```diff
@@ -1751,6 +1751,7 @@ class Log(@volatile var dir: File,
    */
   private def asyncDeleteSegment(segment: LogSegment) {
     segment.changeFileSuffixes("", Log.DeletedFileSuffix)
+
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: Please remove this change.

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 8]
```diff
@@ -230,22 +230,27 @@ class LogManager(logDirs: Seq[File],
...
+        }
+        locks.append(lock)
       } catch {
         case e: IOException =>
+          locks.foreach(_.destroy())
           logDirFailureChannel.maybeAddOfflineLogDir(dir.getAbsolutePath, s"Disk error while locking directory $dir", e)
-          None
       }
     }
+    locks
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogManager.scala:253]: It would be useful to have more context on these changes.

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 9]
```diff
@@ -830,9 +835,11 @@ class LogManager(logDirs: Seq[File],
         cleaner.abortCleaning(topicPartition)
         cleaner.updateCheckpoints(removedLog.dir.getParentFile)
       }
-      removedLog.renameDir(Log.logDeleteDirName(topicPartition))
+
       checkpointLogRecoveryOffsetsInDir(removedLog.dir.getParentFile)
       checkpointLogStartOffsetsInDir(removedLog.dir.getParentFile)
+      removedLog.close()
+      removedLog.renameDir(Log.logDeleteDirName(topicPartition))
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogManager.scala]: Have we verified that it's safe to do the rename after the checkpoint calls? Also, it seems like the following code in `replaceCurrentWithFutureLog` does the `rename` before `close` and needs to be updated too.

```scala
        sourceLog.renameDir(Log.logDeleteDirName(topicPartition))
        // Now that replica in source log directory has been successfully renamed for deletion.
        // Close the log, update checkpoint files, and enqueue this log to be deleted.
        sourceLog.close()
        checkpointLogRecoveryOffsetsInDir(sourceLog.dir.getParentFile)
        checkpointLogStartOffsetsInDir(sourceLog.dir.getParentFile)
```

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. A few comments/questions.

**Comment by ijuma** [comment]
@simplesteph, one more thing: can you please mention the authors of PRs you have used for this? It would be good to include them in the co-authored list.

File: core/src/main/scala/kafka/log/AbstractIndex.scala (hunk 1 above)
Comment (simplesteph) [reply core/src/main/scala/kafka/log/AbstractIndex.scala:362]: to me this is because of https://github.com/apache/kafka/pull/5603/files#diff-ecef8994dd98759052b998e43fffc5b8R277

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 8 above)
Comment (simplesteph) [reply core/src/main/scala/kafka/log/LogManager.scala:253]: @nxmbriggs404 and @GeorgeCGV

**Review by simplesteph — COMMENTED** [review]
@nxmbriggs404 and @GeorgeCGV are the original contributors, I just trimmed, adapted and tested onto the latest trunk. 
I'd be happy to get their inputs

@ijuma I'll incorporate a few suggestions / simplifications and see if it fixes things

**Comment by lindong28** [comment]
It is probably better to include the stacktrace and describe the steps to reproduce the issue in the description of this PR or a new JIRA ticket.

There is following stacktrace provided by @simplesteph in https://github.com/apache/kafka/pull/5491:

```
        Suppressed: java.nio.file.AccessDeniedException: C:\kafka_2.11-2.1.0-SNAPSHOT\data\kafka\second_topic-1 -> C:\kafka_2.11-2.1.0-SNAPSHOT\data\kafka\second_topic-1.d1ceee24d7474152b6fedd61903449e5-delete
                at sun.nio.fs.WindowsException.translateToIOException(Unknown Source)
                at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
                at sun.nio.fs.WindowsFileCopy.move(Unknown Source)
                at sun.nio.fs.WindowsFileSystemProvider.move(Unknown Source)
                at java.nio.file.Files.move(Unknown Source)
                at org.apache.kafka.common.utils.Utils.atomicMoveWithFallback(Utils.java:806)
                ... 23 more
```

This seems to suggest that the user is not authorized to do specific primitive file system operation for this file. If this issue is generally applied to all users who run Kafka on Windows, Kafka won't run on Windows because no segment can be deleted, right? This seems to be different from my impression that the issue happens only occasionally on Windows. So my guess is the stacktrace provided above is due to specific user or system configuration on the machine. It will be good to try reproduce it on another Windows machine that is known to be able to run Kafka most of the time.

**Comment by ijuma** [comment]
Dong, windows doesn't allow certain operations while files are in use. In one cases, we are renaming without closing. In others, we don't unmap, etc.

**Comment by lindong28** [comment]
@ijuma I see. Thank you.

**Comment by HaraldKi** [comment]
I don't think the problem can be solved by point changes in the code. Please follow the links in my comment to the ticket for the likely root cause of all this:

https://issues.apache.org/jira/browse/KAFKA-1194?focusedCommentId=16580774&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16580774

The stack trace you show again is the typical symptom of Java not closing an unmapped file on unmap, but only in a finalizer at some random future time determined by the garbage collector, thereby preventing further access to the file on Windows, in particular, the file cannot be deleted.

**Comment by GeorgeCGV** [comment]
> @nxmbriggs404 and @GeorgeCGV are the original contributors, I just trimmed, adapted and tested onto the latest trunk.
> I'd be happy to get their inputs
> 
> @ijuma I'll incorporate a few suggestions / simplifications and see if it fixes things

The initial contributor will be @nxmbriggs404. I mostly adapted the changes he made to the state of the Kafka base back in the day and verified that it works. 
I don't think I will be much of a help after such a big interval being away from Java|Scala world and Kafka itself.  

> Dong, windows doesn't allow certain operations while files are in use. In one cases, we are renaming without closing. In others, we don't unmap, etc.

Yes, this was the case back then. However, as I remember, the fixes broke other functionality; e.g. transactions.

> I don't think the problem can be solved by point changes in the code. Please follow the links in my comment to the ticket for the likely root cause of all this:
> 
> https://issues.apache.org/jira/browse/KAFKA-1194?focusedCommentId=16580774&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16580774
> 
> The stack trace you show again is the typical symptom of Java not closing an unmapped file on unmap, but only in a finalizer at some random future time determined by the garbage collector, thereby preventing further access to the file on Windows, in particular, the file cannot be deleted.

The fixes did eliminate initial issues on Windows.  In my use case with a single broker, everything worked fine with the changes.
